pyro/infer/trace_mmd.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import absolute_import, division, print_function
+
+from collections import defaultdict
+
+import pyro
+import pyro.ops.jit
+from pyro import poutine
+from pyro.infer.trace_elbo import Trace_ELBO
+from pyro.infer.util import torch_item
+from pyro.util import warn_if_nan
+
+
+def _compute_mmd(X, Z, kernel):
+    mmd = kernel(X).mean() + kernel(Z).mean() - 2 * kernel(X, Z).mean()
+    return mmd
+
+
+class Trace_MMD(Trace_ELBO):
+    """
+    An objective similar to ELBO, but with Maximum Mean Discrepancy (MMD)
+    between marginal variational posterior and prior distributions
+    instead of KL-divergence between variational posterior and prior distributions
+    as in vanilla ELBO. See [1] for the corresponding variant of VAE model,
+    which is a special case of InfoVAE model [2].
+
+    :param kernel: A kernel used to compute MMD.
+        An instance of :class: `pyro.contrib.gp.kernels.kernel.Kernel`,
+        or a dict that maps latent variable names to instances of :class: `pyro.contrib.gp.kernels.kernel.Kernel`.
+        In the latter case, different kernels are used for different latent variables.
+
+    :param mmd_scale: A scaling factor for MMD terms.
```

Comment:
It is ok to handle scaling at initialization (simply with `variance` argument of `Kernel` instance), but I am not sure, if it would be convenient to handle scale change during optimization process via accessing kernel arguments. Also, not every `Kernel` instance has `variance` property.